File: 039--integer-right-triangles/main.py
def triangles(n):
    for a in inclusive_range(1, n - 2):
        for b in inclusive_range(1, a):
            c = n - a - b
            if c > 0:
                yield (a, b, c)
            else:
                break

def inclusive_range(start, stop):
    return range(start, stop + 1)

solution_counts = []
for p in inclusive_range(5, 1000):
    number_of_solutions = 0
    for a, b, c in triangles(p):
        if a*a + b*b == c*c:
            number_of_solutions += 1
    solution_counts.append([p, number_of_solutions])
print(max(solution_counts, key=lambda x: x[1])[0])

File: 039--integer-right-triangles/main.js
function* triangles(n) {
  for (let a = 1; a <= n - 2; a++) {
    for (let b = 1; b <=a; b++) {
      const c = n - a - b;
      if (c > 0) {
        yield [a, b, c];
      } else {
        break;
      }
    }
  }
}

// https://youmightnotneed.com/lodash/#maxBy
const maxBy = (arr, func) => {
  const max = Math.max(...arr.map(func))
  return arr.find(item => func(item) === max)
}

const solutionCounts = [];
for (let p = 5; p <= 1000; p++) {
  let numberOfSolutions = 0;
  for (let [a, b, c] of triangles(p)) {
    if (a*a + b*b === c*c) {
      numberOfSolutions++;
    }
  }
  solutionCounts.push({ p, numberOfSolutions });
}

console.log(maxBy(solutionCounts, x => x.numberOfSolutions).p);
